File: android/lib/daemon-grpc/src/main/kotlin/net/mullvad/mullvadvpn/lib/daemon/grpc/resolver/FixedUdsSocketFactory.java
package net.mullvad.mullvadvpn.lib.daemon.grpc.resolver;

import android.net.LocalSocketAddress;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;

import javax.net.SocketFactory;


class FixedUdsSocketFactory extends SocketFactory {

    private final LocalSocketAddress localSocketAddress;

    public FixedUdsSocketFactory(String path, LocalSocketAddress.Namespace namespace) {
        localSocketAddress = new LocalSocketAddress(path, namespace);
    }

    @Override
    public Socket createSocket() throws IOException {
        return create();
    }

    @Override
    public Socket createSocket(String host, int port) throws IOException {
        return createAndConnect();
    }

    @Override
    public Socket createSocket(String host, int port, InetAddress localHost, int localPort)
            throws IOException {
        return createAndConnect();
    }

    @Override
    public Socket createSocket(InetAddress host, int port) throws IOException {
        return createAndConnect();
    }

    @Override
    public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort)
            throws IOException {
        return createAndConnect();
    }

    private Socket create() {
        return new FixedUdsSocket(localSocketAddress);
    }

    private Socket createAndConnect() throws IOException {
        Socket socket = create();
        SocketAddress unusedAddress = new InetSocketAddress(0);
        socket.connect(unusedAddress);
        return socket;
    }
}

File: android/lib/daemon-grpc/src/main/kotlin/net/mullvad/mullvadvpn/lib/daemon/grpc/resolver/FixedUdsSocketFactory.kt
package net.mullvad.mullvadvpn.lib.daemon.grpc.resolver

import android.net.LocalSocketAddress
import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketAddress
import javax.net.SocketFactory

internal class FixedUdsSocketFactory(path: String?, namespace: LocalSocketAddress.Namespace?) :
    SocketFactory() {
    private val localSocketAddress = LocalSocketAddress(path, namespace)

    @Throws(IOException::class)
    override fun createSocket(): Socket {
        return create()
    }

    @Throws(IOException::class)
    override fun createSocket(host: String, port: Int): Socket {
        return createAndConnect()
    }

    @Throws(IOException::class)
    override fun createSocket(
        host: String,
        port: Int,
        localHost: InetAddress,
        localPort: Int
    ): Socket {
        return createAndConnect()
    }

    @Throws(IOException::class)
    override fun createSocket(host: InetAddress, port: Int): Socket {
        return createAndConnect()
    }

    @Throws(IOException::class)
    override fun createSocket(
        address: InetAddress,
        port: Int,
        localAddress: InetAddress,
        localPort: Int
    ): Socket {
        return createAndConnect()
    }

    private fun create(): Socket {
        return FixedUdsSocket(localSocketAddress)
    }

    @Throws(IOException::class)
    private fun createAndConnect(): Socket {
        val socket = create()
        val unusedAddress: SocketAddress = InetSocketAddress(0)
        socket.connect(unusedAddress)
        return socket
    }
}
